pkg/ingester/ingester.go

```diff
@@ -188,29 +189,28 @@ func (i *Ingester) Push(ctx context.Context, req *connect.Request[pushv1.PushReq
 		level.Debug(instance.logger).Log("msg", "message received by ingester push")
 		for _, series := range req.Msg.Series {
 			for _, sample := range series.Samples {
-				p, size, err := pprof.FromBytes(sample.RawProfile)
-				if err != nil {
-					return nil, err
-				}
-				id, err := uuid.Parse(sample.ID)
-				if err != nil {
-					return nil, err
-				}
-				if err := instance.Head().Ingest(ctx, p, id, series.Labels...); err != nil {
-					reason := validation.ReasonOf(err)
-					if reason != validation.Unknown {
-						validation.DiscardedProfiles.WithLabelValues(string(reason), instance.tenantID).Add(float64(1))
-						validation.DiscardedBytes.WithLabelValues(string(reason), instance.tenantID).Add(float64(size))
-						switch validation.ReasonOf(err) {
+				err := pprof.FromBytes(sample.RawProfile, func(p *profilev1.Profile, size int) error {
+					id, err := uuid.Parse(sample.ID)
+					if err != nil {
+						return err
+					}
+					if err = instance.Head().Ingest(ctx, p, id, series.Labels...); err != nil {
+						switch reason := validation.ReasonOf(err); reason {
 						case validation.OutOfOrder:
-							return nil, connect.NewError(connect.CodeInvalidArgument, err)
+							return connect.NewError(connect.CodeInvalidArgument, err)
 						case validation.SeriesLimit:
-							return nil, connect.NewError(connect.CodeResourceExhausted, err)
+							return connect.NewError(connect.CodeResourceExhausted, err)
+						default:
+							validation.DiscardedProfiles.WithLabelValues(string(reason), instance.tenantID).Add(float64(1))
+							validation.DiscardedBytes.WithLabelValues(string(reason), instance.tenantID).Add(float64(size))
+							return err
 						}
 					}
+					return nil
+				})
+				if err != nil {
 					return nil, err
 				}
```

Comment:
I'm not sure if we need to break the loop on the first error encountered. If we do need, then I guess we should bump the `DiscardedProfiles` metrics by the number of profiles we actually dropped, like `len(series.Samples)-i`. However, it does not work for `DiscardedBytes`

Reply:
Yes I fully agree, we should ingest all the profiles that are valid (== soft error), and then returning the error appropriate for the offending sample. Otherwise we would loose perfectly valid samples. 

This is how mimir is doing it: https://github.com/grafana/mimir/blob/main/pkg/ingester/ingester.go#L826-L828

Reply:
I agree the scope of that is beyond this PR.